src/lang-promql/parser/parser.ts

```diff
@@ -273,6 +265,46 @@ export class Parser {
     }
   }
 
+  private checkVectorSelector(node: Subtree): void {
+    const labelMatchers = buildLabelMatchers(
+      retrieveAllRecursiveNodes(walkThrough(node, LabelMatchers, LabelMatchList), LabelMatchList, LabelMatcher),
+      this.state
+    );
+    let vectorSelectorName = '';
+    // VectorSelector ( MetricIdentifier ( Identifier ) )
+    // https://github.com/promlabs/lezer-promql/blob/71e2f9fa5ae6f5c5547d5738966cd2512e6b99a8/src/promql.grammar#L200
+    const vectorSelectorNodeName = node.firstChild?.firstChild;
+    if (vectorSelectorNodeName) {
+      vectorSelectorName = this.state.sliceDoc(vectorSelectorNodeName.start, vectorSelectorNodeName.end);
+    }
+    if (vectorSelectorName !== '') {
+      labelMatchers.push(new Matcher(EqlSingle, '__name__', vectorSelectorName));
+    }
+    console.log(`${this.tree}`);
+    console.log(labelMatchers);
+    // A Vector selector must contain at least one non-empty matcher to prevent
+    // implicit selection of all metrics (e.g. by a typo).
+    const empty = labelMatchers.every((lm) => lm.matchesEmpty());
+    if (empty) {
+      this.addDiagnostic(node, 'vector selector must contain at least one non-empty matcher');
+    }
+
+    if (vectorSelectorName !== '') {
+      // In this case the last LabelMatcher is checking for the metric name
+      // set outside the braces. This checks if the name has already been set
+      // previously
+      let i = 0;
+      let metricNameSet = false;
+      while (i < labelMatchers.length - 1 && !metricNameSet) {
+        metricNameSet = labelMatchers[i].name === '__name__';
+        i++;
+      }
+      if (metricNameSet) {
+        this.addDiagnostic(node, `metric name must not be set twice: ${vectorSelectorName} or ${labelMatchers[i - 1].value}`);
```

Comment:
Ah because of your `i` here... though you could then do:

```suggestion
        this.addDiagnostic(node, `metric name must not be set twice: ${vectorSelectorName} or ${labelMatchers.filter(lm -> lm.name === '__name__')[0].value}`);
```

Reply:
Even better, do this above:

```typescript
const labelMatcherMetricName = labelMatchers.find(lm => lm.name === '__name__');
```

That returns `undefined` in case there is no metric name.

Reply:
And then `labelMatcherMetricName.value` here.

Reply:
If I'm doing that it will find the LabelMatcher I manually inject in the array.

Reply:
Maybe just do this first then, before extending the array :)

Reply:
yup I did it like you suggested, much cleaner. Thanks !